src/udev/udev-watch.c

```diff
@@ -161,36 +165,450 @@ static int synthesize_change_one(sd_device *dev, sd_device *target) {
+                .uuid = TAKE_PTR(uuid),
+        };
+
+        r = hashmap_ensure_put(partitions, &partition_hash_ops, p->node, p);
+        if (r < 0)
+                return r;
+
+        TAKE_PTR(p);
+        return 0;
+}
+
+static int enumerate_partitions_by_blkid(sd_device *dev, Hashmap **ret) {
+        int r;
+
+        assert(dev);
+        assert(ret);
 
-        /* search for partitions */
+        _cleanup_close_ int fd = sd_device_open(dev, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
+        if (fd < 0)
+                return fd;
+
+        if (flock(fd, LOCK_EX|LOCK_NB) < 0)
+                return -errno;
+
+        r = dlopen_libblkid();
+        if (r < 0)
+                return r;
+
+        _cleanup_(blkid_free_probep) blkid_probe b = sym_blkid_new_probe();
+        if (!b)
+                return -ENOMEM;
+
+        errno = 0;
+        r = sym_blkid_probe_set_device(b, fd, 0, 0);
+        if (r != 0)
+                return errno_or_else(ENOMEM);
+
+        sym_blkid_probe_enable_partitions(b, 1);
+        sym_blkid_probe_set_partitions_flags(b, BLKID_PARTS_ENTRY_DETAILS);
+
+        errno = 0;
+        r = sym_blkid_do_safeprobe(b);
+        if (r == _BLKID_SAFEPROBE_ERROR)
+                return errno_or_else(EIO);
+        if (IN_SET(r, _BLKID_SAFEPROBE_AMBIGUOUS, _BLKID_SAFEPROBE_NOT_FOUND)) {
+                /* Maybe broken partition table or uninitialized disk. Anyway, no valid partition should
+                 * exist. */
+                *ret = NULL;
+                return 0;
+        }
+        assert(r == _BLKID_SAFEPROBE_FOUND);
+
+        errno = 0;
+        blkid_partlist pl = sym_blkid_probe_get_partitions(b);
+        if (!pl) {
+                if (errno > 0)
+                        return -errno;
+
+                /* Maybe the whole disk is used as a filesystem or crypt device. Anyway, no valid partition
+                 * should exist.*/
+                *ret = NULL;
+                return 0;
+        }
+
+        errno = 0;
+        int n_partitions = sym_blkid_partlist_numof_partitions(pl);
+        if (n_partitions < 0)
+                return errno_or_else(EIO);
+
+        _cleanup_hashmap_free_ Hashmap *partitions = NULL;
+        for (int i = 0; i < n_partitions; i++) {
+                blkid_partition pp;
+
+                errno = 0;
+                pp = sym_blkid_partlist_get_partition(pl, i);
+                if (!pp)
+                        return errno_or_else(EIO);
+
+                r = partition_add(dev, pp, &partitions);
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(partitions);
+        return 0;
+}
+
+static int partitions_synced(sd_device *dev) {
+        int r;
+
+        assert(dev);
+
+        /* search for partitions by blkid */
+        _cleanup_hashmap_free_ Hashmap *partitions = NULL;
+        r = enumerate_partitions_by_blkid(dev, &partitions);
+        if (r < 0)
+                return r;
```

Comment:
The result of this should be the same every iteration right? Shouldn't we calculate it once and cache it in SyncDeviceContext?

Reply:
If not please add a comment why this would change over consecutive iterations.